common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+        m_address.sll_halen   = sizeof(sMacAddr);
+        std::copy_n(dst_mac.oct, sizeof(sMacAddr), m_address.sll_addr);
+
+        m_length = sizeof(m_address);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_ll m_address = {};
+    socklen_t m_length;
+};
+
+class NetlinkAddress : public SockAddr {
+public:
+    NetlinkAddress(uint32_t groups)
+    {
+        // TODO
+        m_address.nl_family = AF_NETLINK;
+        m_address.nl_groups = groups;
+
+        m_length = sizeof(m_address);
+    }
+
+    const struct sockaddr *sockaddr() const override
+    {
+        return reinterpret_cast<const struct sockaddr *>(&m_address);
+    }
+    const socklen_t &length() const override { return m_length; }
+
+private:
+    sockaddr_nl m_address = {};
+    socklen_t m_length;
+};
+
+/**
+ * This class is a wrapper for the socket file descriptor obtained with the accept() system call.
+ */
+class ConnectedSocket : public SocketAbstractImpl {
+public:
+    explicit ConnectedSocket(int fd) : SocketAbstractImpl(fd) {}
+};
+
+class RawSocket : public SocketAbstractImpl {
+public:
+    explicit RawSocket(uint16_t protocol = ETH_P_ALL)
+        : SocketAbstractImpl(socket(AF_PACKET, SOCK_RAW, htons(protocol)))
+    {
+    }
+};
+
+class UdpSocket : public SocketAbstractImpl {
+public:
+    UdpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)) {}
+};
+
+class TcpSocket : public SocketAbstractImpl {
+public:
+    TcpSocket() : SocketAbstractImpl(socket(AF_INET, SOCK_STREAM, IPPROTO_IP)) {}
+};
+
+class UdsSocket : public SocketAbstractImpl {
+public:
+    explicit UdsSocket() : SocketAbstractImpl(socket(AF_UNIX, SOCK_STREAM, 0)) {}
+};
+
+class AbstractNetlinkSocket : public SocketAbstractImpl {
+protected:
+    explicit AbstractNetlinkSocket(uint16_t protocol)
+        : SocketAbstractImpl(socket(AF_NETLINK, SOCK_RAW, protocol))
+    {
+    }
+};
+
+class NetlinkRouteSocket : public AbstractNetlinkSocket {
+public:
+    NetlinkRouteSocket() : AbstractNetlinkSocket(NETLINK_ROUTE) {}
+};
+
+/**
+ * This class implements the Connection interface with methods that wrap the recv(), recvfrom and
+ * recvmsg() system calls.
+ */
+class SocketConnectionImpl : public Connection {
+public:
+    explicit SocketConnectionImpl(const std::shared_ptr<Socket> &socket) : m_socket(socket) {}
+
+    std::shared_ptr<Socket> socket() override { return m_socket; }
+    int receive(Buffer &buffer) override
+    {
+        return ::recv(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT);
+    }
+    int receive_from(Buffer &buffer, SockAddr &address) override
+    {
+        return ::recvfrom(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT,
+                          address.sockaddr(), &address.length());
+    }

```

Comment:
As mentioned, addrlen is an inout argument, so we have to give it a real modifiable value:
```suggestion
    int receive_from(Buffer &buffer, SockAddr &address) override
    {
        size_t addrlen = address.length();
        return ::recvfrom(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT,
                          address.sockaddr(), &addrlen);
    }
```

Note that all cases we support have fixed address length. We could check that `addrlen == address.length` afterwards, but there's not really any need.

More importantly though, we should make sure that the SockAddr passed into receive_from corresponds exactly to the sockaddr type corresponding to the address family of the socket. So I think we have to unify the SockAddr and Socket subclasses into a single specialized Connection class per connection type.

Note that the raw socket is the only one where recvfrom makes sense. All the others are typically bind()'ed to something. Well, you can make a UDP server without bind/listen/accept but for us there's no reason to do that (it's typical for e.g. NTP, DNS, ... which are single-request-response "connections").

Reply:
I agree with suggested change about address length.
I see the problem you are describing (it's also happening with the underlying C system calls and some simple C++ wrappers will not fix it). I don't see however how to unify SockAddr and Socket subclasses.

Reply:
Actually I think there is no point to have a SockAddr class, it's better if the specialized classes expose a specialized function. For example, RawSocket would have `receive_from(Buffer& buffer, sMacAddr& src)`.